# Calcul Numeric - Subiecte de examen

**Notă**: formulele de $\LaTeX$ nu se văd prea bine în viewer-ul de pe GitHub, e mai bine să descărcați acest document local.

In [1]:
import math
import numpy as np
import scipy.spatial
import matplotlib.pyplot as plt

## Exercițiul 1

Metoda bisecției - secțiunea I.1 din [cursul 1](https://drive.google.com/file/d/1hUFKessAxlGIT_gNFlH7-4yyuyQ_4vFe/view)

### a)

Pentru a găsi intervalele de monotonie ale lui $f(x) = x^3 - 4x + 1$, calculăm derivata funcției: $f'(x) = 3x^2 - 4$.

Vedem că derivata se anulează în $\pm \frac{2 \sqrt{3}}{3}$. Deci intervalele de monotonie sunt:
- $\left(-\infty, -\frac{2 \sqrt{3}}{3}\right)$
- $\left(-\frac{2 \sqrt{3}}{3}, +\frac{2 \sqrt{3}}{3}\right)$
- $\left(+\frac{2 \sqrt{3}}{3}, +\infty\right)$

### b)

Observăm că $f\left(+ \frac{2 \sqrt{3}}{3}\right) < 0$, și de exemplu $f(4) > 0$. Deci în acest interval funcția are cel puțin o soluție reală. Dar deoarece este și monotonă pe $\left(\frac{2 \sqrt{3}}{3}, +\infty\right)$, această soluție este unică.

### c)

In [2]:
def bisection_search(f, a, b, epsilon=1e-3):
    # Calculăm valorile în capete
    f_a = f(a)
    f_b = f(b)

    # Prima estimare, mijlocul intervalului inițial
    x_num = (a + b) / 2

    # Numărul necesar de iterații
    num_iterations = math.floor(math.log2((b - a) / epsilon) - 1) + 1
    
    last_rel_error = math.inf

    # Aplicăm algoritmul
    for step in range(num_iterations):
    
        print(f"a = {a}, b = {b}, error = {last_rel_error}")
        value = f(x_num)

        # Am găsit fix valoarea căutată, ieșim
        if value == 0:
            break
        elif f_a * value < 0:
            b = x_num
        else:
            a = x_num
        
        x_num_vechi = x_num
        x_num = (a + b) / 2
        
        last_rel_error = abs((x_num - x_num_vechi)) / abs(x_num_vechi + epsilon)

    return x_num

f = lambda x: x ** 3 - 4 * x + 1
print("x1 =", bisection_search(f, -4, -1))
print("x2 =", bisection_search(f, -1, 1))
print("x3 =", bisection_search(f, 1, 4))

a = -4, b = -1, error = inf
a = -2.5, b = -1, error = 0.30012004801920766
a = -2.5, b = -1.75, error = 0.2144082332761578
a = -2.125, b = -1.75, error = 0.08827683615819208
a = -2.125, b = -1.9375, error = 0.04841208365608055
a = -2.125, b = -2.03125, error = 0.023088289619504987
a = -2.125, b = -2.078125, error = 0.011283625203105253
a = -2.125, b = -2.1015625, error = 0.005578862804605909
a = -2.125, b = -2.11328125, error = 0.0027739558830056367
a = -2.119140625, b = -2.11328125, error = 0.0013831411689202646
a = -2.1162109375, b = -2.11328125, error = 0.0006925284490686877
x1 = -2.115478515625
a = -1, b = 1, error = inf
a = 0.0, b = 1, error = 500.0
a = 0.0, b = 0.5, error = 0.499001996007984
a = 0.25, b = 0.5, error = 0.49800796812749004
a = 0.25, b = 0.375, error = 0.16622340425531915
a = 0.25, b = 0.3125, error = 0.09968102073365231
a = 0.25, b = 0.28125, error = 0.05535872453498671
a = 0.25, b = 0.265625, error = 0.029301453352086265
a = 0.25, b = 0.2578125, error = 0.015092972

## Exercițiul 2

Metoda Newton-Rhapson - secțiunea I.2 din [cursul 1](https://drive.google.com/file/d/1hUFKessAxlGIT_gNFlH7-4yyuyQ_4vFe/view)

### a)

#### Formula

Formula pentru următorul din șirul aproximărilor este

$$x_k = x_{k-1} - \frac{f (x_{k-1})}{f'(x_{k - 1})}$$

#### Convergență 

Fie $f \in C^2([a, b])$, $f'$ și $f''$ nu se anulează pe $[a, b]$ și $f(a) f(b) < 0$. 

Fie $x_0 \in [a, b]$ astfel încât să aibă loc condiția

$$f(x_0) f''(x_0) > 0$$

atunci ecuația $f(x) = 0$ are o soluție unică $x^* \in (a, b)$ iar șirul $x_k$ construit prin metoda Newton-Rhapson converge la $x^*$.

### b)

Fie $f(x) = x^2 + 2 x − 1$, $f \colon [0.1, 2] \to \mathbb{R}$.

Atunci $f'(x) = 2x + 2$, $f''(x) = 2$.

Avem că:
- $f$ este de clasă $C^2$
- $f'$ și $f''$ nu se anulează pe $[0.1, 2]$
- $f(0.1) f(2) = -0.79 \cdot 7 < 0$

Dacă luăm $x_0 = 1$ avem $f(x_0) f''(x_0) = 2 \cdot 2 > 0$, deci ipotezele teoremei de convergență sunt satisfăcute.

### c)

In [3]:
def newton_rhapson(f, df, x0, epsilon=1e-3):
    # Primul punct este cel primit ca parametru
    prev_x = x0
    # Aplicăm prima iterație
    x = x0 - f(x0) / df(x0)

    # Continuăm să calculăm până avem precizia cerută.
    last_rel_error = math.inf
    while abs(x - prev_x) / abs(prev_x) > epsilon:
        print(f"x = {x}, error = {last_rel_error}")
        x, prev_x = x - f(x) / df(x), x

        last_rel_error = abs((x - prev_x)) / abs(prev_x)

    return x

f = lambda x: x**2 + 2 * x - 1
df = lambda x: 2 * x + 2
x0 = 1
newton_rhapson(f, df, x0)

x = 0.5, error = inf
x = 0.4166666666666667, error = 0.16666666666666663
x = 0.41421568627450983, error = 0.00588235294117645


0.41421356237468987

## Exercițiul 3

Metoda lui Gauss - [cursul 3](https://drive.google.com/file/d/1huTen5Elw203cEOoQMRt1gchFciXlIXp/view)

In [4]:
A1 = np.array([
    [1, 0, -2],
    [2, -2, 0],
    [1, 2, 1],
], dtype=np.float64)
b1 = np.array([
    [-5],
    [-6],
    [5]
], dtype=np.float64)

A2 = np.array([
    [2, -2, 1],
    [1, 3, -2],
    [3, -1, -1]
], dtype=np.float64)
b2 = np.array([
    [-3],
    [1],
    [2]
], dtype=np.float64)

In [5]:
def compute_upper_triangular(A, b, partial_pivot=False, full_pivot=False):
    """Aduce un sistem la forma superior triunghiulară
    folosind Gauss cu/fără pivotare totală/parțială.
    """
    if partial_pivot and full_pivot:
        raise Exception("Se poate folosi doar un tip de pivotare")

    N = A.shape[0]

    # Formez matricea sistemului
    M = np.concatenate((A, b), axis=1)

    indices = np.arange(0, N)

    for k in range(N - 1):
        print(M)

        if partial_pivot:
            # Găsesc indicele elementului de valoare maximă
            index = np.argmax(M[k:, k])

            # Pivotez
            M[[k, index]] = M[[index, k]]
        elif full_pivot:
            submatrix = M[k:, k:N - 1]
            index = np.unravel_index(np.argmax(np.abs(submatrix)), submatrix.shape)

            # Obțin indicii în matricea mare
            index = (index[0] + k, index[1] + k)

            M[[k, index[0]]] = M[[index[0], k]]
            M[:, [k, index[1]]] = M[:, [index[1], k]]

            indices[[k, index[1]]] = indices[[index[1], k]]

        # Selectez coloana
        ratios = M[k + 1:, k]

        # Determin raportul pentru fiecare rând
        ratios = ratios / M[k, k]

        row = M[k, :]

        # Înmulțesc fiecare raport cu primul rând
        difference = np.outer(ratios, row)

        # Actualizez matricea
        M[k + 1:, :] -= difference

    return M, indices


def solve_upper_triangular(U, C, print_steps=False):
    """Rezolvă un sistem în formă superior triunghiulară,
    folosind substituția descendentă.
    """
    N = U.shape[0]

    x = np.zeros(N)

    # Merg de la ultima linie către prima,
    # și rezolv pe rând ecuațiile prin substituție
    for i in range(N - 1, -1, -1):
        coefs = U[i, i + 1:]
        values = x[i + 1:]

        x[i] = (C[i] - coefs @ values) / U[i, i]

        if print_steps:
            print("Pasul", N - i, ":",
                C[i], "-", coefs, "@", values, ")",
                "/", U[i, i])

    return x


def solve_system(A, b, pivot=None):
    "Rezolvă un sistem folosind metoda Gauss."

    determinant = np.linalg.det(A)

    if np.abs(determinant) < 1e-14:
        # Sistemul nu are soluție
        return None

    partial_pivot = False
    full_pivot = False

    if pivot == "full":
        full_pivot = True
    elif pivot == "partial":
        partial_pivot = True

    M, indices = compute_upper_triangular(A, b, partial_pivot=partial_pivot, full_pivot=full_pivot)

    N = A.shape[0]
    U = M[:,:N]
    C = M[:,N]

    x = solve_upper_triangular(U, C)
    x = x[indices]

    return x

print(solve_system(A1, b1, pivot='partial'))
print()
print(solve_system(A2, b2, pivot='full'))

[[ 1.  0. -2. -5.]
 [ 2. -2.  0. -6.]
 [ 1.  2.  1.  5.]]
[[ 2. -2.  0. -6.]
 [ 0.  1. -2. -2.]
 [ 0.  3.  1.  8.]]
[-1.  2.  2.]

[[ 2. -2.  1. -3.]
 [ 1.  3. -2.  1.]
 [ 3. -1. -1.  2.]]
[[ 3.          1.         -2.          1.        ]
 [ 0.          2.66666667 -0.33333333 -2.33333333]
 [ 0.          3.33333333 -1.66666667  2.33333333]]
[-1.4 -2.  -4.2]


## Exercițiul 4

Descompunerea LU - [cursul 4](https://drive.google.com/file/d/14jt-8XcekhX8CgMgMgtnwtDEpr_lzd02/view)

In [6]:
A = np.array([
    [2, 3, 1],
    [4, 8, 4],
    [-4, 0, 10],
], dtype=np.float64)


b = np.array([
    [5],
    [18],
    [20],
], dtype=np.float64)

In [7]:

N = A.shape[0]
P = np.eye(N)
L = np.zeros((N, N))
U = np.copy(A)

partial_pivot = True

for k in range(N - 1):
    if partial_pivot:
        # Găsesc indicele elementului de magnitudine maximă
        index = k + np.argmax(np.abs(U[k:, k]))

        # Pivotez
        U[[k, index]] = U[[index, k]]
        L[[k, index]] = L[[index, k]]

        # Interschimb în permutare
        P[[k, index]] = P[[index, k]]

    # Selectez coloana pe care lucrez
    ratios = U[k + 1:, k]

    # Determin raportul pentru fiecare rând
    ratios = ratios / U[k, k]

    # Actualizez matricea inferior triunghiulară
    L[k + 1:, k] = ratios

    # Selectez rândul pe care vreau să-l actualizez
    row = U[k, :]

    # Înmulțesc fiecare raport cu primul rând
    difference = np.outer(ratios, row)

    # Actualizez matricea superior triunghiulară
    U[k + 1:, :] -= difference

L += np.eye(N)

print("L = ")
print(L)
print("U = ")
print(U)
print("L @ U = ")
print(L @ U)
print("P @ A = ")
print(P @ A)
print()


print("Rezolv pentru b = ", b.T)

# Permut numerele din vector
b = P @ b


y = np.zeros(N)

# Merg de la prima linie în jos,
# și rezolv pe rând ecuațiile prin substituție
for i in range(0, N):
    coefs = L[i, :i + 1]
    values = y[:i + 1]

    y[i] = (b[i] - coefs @ values) / L[i, i]

print("Obțin y = ", y)


x = np.zeros(N)

# Merg de la ultima linie în sus,
# și rezolv pe rând ecuațiile prin substituție
for i in range(N - 1, -1, -1):
    coefs = U[i, i + 1:]
    values = x[i + 1:]

    x[i] = (y[i] - coefs @ values) / U[i, i]

print("Obțin x = ", x)
print("Verificare: A @ x = ", A @ x)

L = 
[[ 1.     0.     0.   ]
 [-1.     1.     0.   ]
 [ 0.5   -0.125  1.   ]]
U = 
[[ 4.    8.    4.  ]
 [ 0.    8.   14.  ]
 [ 0.    0.    0.75]]
L @ U = 
[[ 4.  8.  4.]
 [-4.  0. 10.]
 [ 2.  3.  1.]]
P @ A = 
[[ 4.  8.  4.]
 [-4.  0. 10.]
 [ 2.  3.  1.]]

Rezolv pentru b =  [[ 5. 18. 20.]]
Obțin y =  [18.   38.    0.75]
Obțin x =  [-2.5  3.   1. ]
Verificare: A @ x =  [ 5. 18. 20.]


## Exercițiul 5

In [8]:
A = np.array([
    [4, 2, 4],
    [2, 2, 3],
    [4, 3, 14]
])

### a)

In [9]:
def simetrica(M):
    "Verifică dacă matricea `M` este simetrică."
    return np.all(M == M.T)

simetrica(A)

True

In [10]:
def pozitiv_semidefinita(M):
    "Verifică dacă matricea M este pozitiv-semidefinită."
    for i in range(1, M.shape[0]+1):
        minor = M[:i, :i]
        # Dacă cel puțin un minor principal nu are determinantul nenegativ,
        # nu este pozitiv-semidefinită.
        if np.linalg.det(minor) < 0:
            return False
    return True

pozitiv_semidefinita(A)

True

### b)

In [11]:
def descompunere_cholesky(M):
    N = M.shape[0]

    L = np.eye(N)

    for i in range(N):
        L_i = np.eye(N)

        pivot = M[i, i]
        L_i[i:, i] = M[i:, i] / np.sqrt(pivot)

        M_nou = np.eye(N)
        outer = np.outer(M[i + 1:, i], M[i, i + 1:])
        M_nou[i + 1:, i + 1:] = M[i + 1:, i + 1:] - outer / pivot

        L = L @ L_i
        M = M_nou

    return L

L = descompunere_cholesky(A)
print(L)

[[2. 0. 0.]
 [1. 1. 0.]
 [2. 1. 3.]]


### c)

Din descompunerea Cholesky am obținut $L$, și știm că $A = L L^T$.

Ecuația se rescrie

$$L L^T x = b$$

Deci putem rezolva mai întâi sistemul inferior triunghiular

$$L y = b$$

Și apoi cel superior triunghiular

$$L^T x = y$$

In [12]:
b = np.array([
    [10],
    [6],
    [11]
])

In [13]:
def substitutie_descendenta(U, b):
    N = b.shape[0]
    x = np.zeros(N)

    for i in range(N - 1, -1, -1):
        coefs = U[i, i + 1:]
        values = x[i + 1:]

        x[i] = (b[i] - coefs @ values) / U[i, i]

    return x

def substitutie_ascendenta(L, b):
    N = b.shape[0]
    x = np.zeros(N)

    for i in range(0, N):
        coefs = L[i, :i + 1]
        values = x[:i + 1]

        x[i] = (b[i] - coefs @ values) / L[i, i]

    return x

y = substitutie_ascendenta(L, b)
x = substitutie_descendenta(L.T, y)
print("x =",x)

x = [2. 1. 0.]


## Exercițiul 6

Metoda Jacobi - [cursul 7](https://drive.google.com/file/d/1Dk_6UFiM674Zwdm6VRAc9-a1rVQAH7Mu/view) sau [pe Wikipedia](https://en.wikipedia.org/wiki/Jacobi_eigenvalue_algorithm)

In [14]:
A = np.array([
    [4, 1, 1],
    [1, 4, 1],
    [1, 1, 4]
], dtype=np.float64)

In [15]:
def gaseste_max_nediagonal(M):
    "Caută elementul de valoare absolută maximă care nu se află pe diagonala principală"
    
    # Matricea e simetrică, suficient să caut sub diagonala principală
    sub_diagonala = np.tril(M, k=-1)

    # Caut elementul de magnitudine maximă
    index = np.argmax(np.abs(sub_diagonala))

    # Returnez indicii elementului
    return np.unravel_index(index, M.shape)

def construieste_matrice_de_rotatie(i, j, theta):
    direction = np.zeros(3)
    direction[i] = 1
    direction[j] = 1

    r = scipy.spatial.transform.Rotation.from_rotvec(theta * direction)
    
    return r.as_matrix()

def pas_metoda_jacobi(M):
    "Aplică o iterație din metoda Jacobi"
    
    p, q = gaseste_max_nediagonal(M)
    
    if np.abs(M[p, p] - M[q, q]) < 1e-7:
        theta = np.pi / 4
    else:
        theta = np.arctan(2 * M[p, q] / (M[q, q] - M[p, p]))/2

    R = construieste_matrice_de_rotatie(p, q, theta)
    
    return R @ M @ R.T

print("Matricea inițială:")
print(A)
print()

A1 = pas_metoda_jacobi(A)
print("După primul pas:")
print(A1)
print()

A2 = pas_metoda_jacobi(A1)
print("După al doilea pas:")
print(A2)
print()

print("Valorile proprii aproximative:")
print(np.diagonal(A2))

Matricea inițială:
[[4. 1. 1.]
 [1. 4. 1.]
 [1. 1. 4.]]

După primul pas:
[[5.6685871  0.59857503 0.72533587]
 [0.59857503 3.13426284 0.16269581]
 [0.72533587 0.16269581 3.19715007]]

După al doilea pas:
[[5.24160664 0.97206464 0.86897061]
 [0.97206464 3.42153233 0.37682597]
 [0.86897061 0.37682597 3.33686103]]

Valorile proprii aproximative:
[5.24160664 3.42153233 3.33686103]


## Exercițiul 7

Polinom de interpolare Lagrange - [cursul 8](https://drive.google.com/file/d/14YiX_LDd2cOAiLBF4dyHlWcWzqRZ8htm/view)

Fie o funcție $f$. Avem nodurile de interpolare $x_1$, $x_2$, $x_3$ și valorile corespunzătoare $y_1 = f(x_1)$, $y_2 = f(x_2)$, $y_3 = f(x_3)$.


### a) Metoda directă

Construim și rezolvăm sistemul
$$
\begin{pmatrix}
1 & x_1 & x_1^2 \\
1 & x_2 & x_2^2 \\
1 & x_3 & x_3^2
\end{pmatrix}
\cdot
\begin{pmatrix}
a_1 \\
a_2 \\
a_3
\end{pmatrix}
=
\begin{pmatrix}
y_1 \\
y_2 \\
y_3
\end{pmatrix}
$$
unde $a_1$, $a_2$, $a_3$ sunt coeficienții polinomului cerut
$$
P_2(x) = a_1 + a_2 x + a_3 x^2
$$

### b) Metoda Lagrange

Construim polinoamele $L_{n, k}$:
$$
L_{2, 1}(x) = \frac{(x - x_2) (x - x_3)}{(x_1 - x_2) (x_1 - x_3)} \\
L_{2, 2}(x) = \frac{(x - x_1) (x - x_3)}{(x_2 - x_1) (x_2 - x_3)} \\
L_{2, 3}(x) = \frac{(x - x_1) (x - x_2)}{(x_3 - x_1) (x_3 - x_2)}
$$

Polinomul cerut este
$$
P_2(x) = L_{2, 1}(x) \cdot y_1 + L_{2, 2}(x) \cdot y_2 + L_{2, 3}(x) \cdot y_3
$$

### c) Metoda Newton

Construim și rezolvăm sistemul inferior triunghiular
$$
\begin{array}{lcl}
c_1 & = & y_1 \\
c_1 + c_2 (x_2 - x_1) & = & y_2 \\
c_1 + c_2 (x_3 - x_1) + c_3 (x_3 - x_1) (x_3 - x_2) & = & y_3
\end{array}
$$

Polinomul cerut este
$$
P_2(x) = c_1 + c_2 (x - x_1) + c_3 (x - x_1) (x - x_2)
$$

### d) Metoda Newton cu diferențe divizate

Diferențele divizate de ordin 0, 1, 2 sunt:
$$
f[x_1] = f(x_1) \\
f[x_1, x_2] = \frac{f[x_2] - f[x_1]}{x_2 - x_1} \\
f[x_1, x_2, x_3] = \frac{f[x_2, x_3] - f[x_1, x_2]}{x_3 - x_1}
$$

Construim un tabel cu diferențele pentru cazul nostru:

| $x_i$ | DD ordin 0 | DD ordin 1 | DD ordin 2 |
|-------|------------|------------|------------|
| $x_1$ | $$f[x_1] = f(x_1) = y_1$$ | | |
| $x_2$ | $$f[x_2] = f(x_2) = y_2$$ | $$f[x_1, x_2]$$ | |
| $x_3$ | $$f[x_3] = f(x_3) = y_3$$ | $$f[x_2, x_3]$$ | $$f[x_1, x_2, x_3]$$ |

Polinomul cerut este
$$
P_2(x) = f[x_1] + f[x_1, x_2] (x - x_1) + f[x_1, x_2, x_3] (x - x_1) (x - x_2)
$$

## Exercițiul 8

### a)

$$
f(x) = \ln x \\
x = (1, e, e^2)
$$

Cu metoda Lagrange obținem:

$$
L_{2, 1}(x) = \frac{(x - e) (x - e^2)}{(1 - e) (1 - e^2)} \\
L_{2, 2}(x) = \frac{(x - 1) (x - e^2)}{(e - 1) (e - e^2)} \\
L_{2, 3}(x) = \frac{(x - 1) (x - e)}{(e^2 - 1) (e^2 - e)} \\
P_2(x) =  L_{2, 1}(x) \cdot \ln 1 + L_{2, 2}(x) \cdot \ln e + L_{2, 3}(x) \cdot \ln e^2
$$

Cu metoda Newton cu diferențe divizate obținem:

$$
f[x_1] = \ln 1; f[x_2] = \ln e; f[x_3] = \ln e^2 \\
f[x_1, x_2] = \frac{\ln e - \ln 1}{e - 1}; f[x_2, x_3] = \frac{\ln e^2 - \ln e}{e^2 - e} \\
f[x_1, x_2, x_3] = \frac{f[x_2, x_3] - f[x_1, x_2]}{e^2 - 1} \\
P_2(x) = f[x_1] + f[x_1, x_2] (x - 1) + f[x_1, x_2, x_3] (x - 1) (x - e)
$$

### b)

Analog cu cel de mai sus, dar pentru $f(x) = \sin \left(\frac{\pi x}{2}\right)$ și $x = (-1, 0, 1)$.

## Exercițiul 9

Avem punctele $(-2, 1)$, $(-1, 4)$, $(0, 11)$, $(1, 16)$, $(2, 13)$, $(3, -4)$.

Calculăm tabelul diferențelor divizate:


| $x_i$ | DD ordin 0 | DD ordin 1 | DD ordin 2 | DD ordin 3 | DD ordin 4 | DD ordin 5 |
|-------|------------|------------|------------|------------|------------|------------|
| $$-2$$ | $$1$$ |
| $$-1$$ | $$4$$ | $$3$$ |
| $$0$$ | $$11$$ | $$7$$ | $$2$$ |
| $$1$$ | $$16$$ | $$5$$ | $$-1$$ | $$-1$$ |
| $$2$$ | $$13$$ | $$-3$$ | $$-4$$ | $$-1$$ | $$0$$ |
| $$3$$ | $$-4$$ | $$-17$$ | $$-7$$ | $$-1$$ | $$0$$ | $$0$$ |

Observăm că diferențele divizate de ordin 4 sau mai mare sunt 0, deci polinomul pe care l-am obține e de grad 3.

## Exercițiul 10

Avem punctele $(0, 0)$, $(0.5, y)$, $(1, 3)$, $(2, 2)$.

Polinomul de interpolare obținut prin metoda Lagrange este

$$
\begin{align}
P_3(x) &= L_{3, 1}(x) y_1 + L_{3, 2}(x) y_2 + L_{3, 3}(x) y_3 + L_{3, 4}(x) y_4 \\
&= 0 + L_{3, 2}(x) y + L_{3, 3}(x) \cdot 3 + L_{3, 4}(x) \cdot 2 \\
&= \frac{(x - 0)(x - 1)(x - 2)}{(0.5 - 0) (0.5 - 1) (0.5 - 2)} \cdot y + \frac{(x - 0)(x - 0.5)(x - 2)}{(1 - 0)(1 - 0.5)(1 - 2)} \cdot 3 + \frac{(x - 0)(x - 0.5) (x - 1)}{(2 - 0)(2 - 0.5)(2 - 1)} \cdot 2
\end{align}
$$

Dacă scoatem doar termenii de grad 3 rămânem cu

$$
\frac{x^3}{0.375} \cdot y + \frac{x^3}{-0.5} \cdot 3 + \frac{x^3}{3} \cdot 2
$$

Din cerință, știm că coeficientul acestui termen trebuie să fie egal cu 6, deci

$$
6 = \frac{y}{0.375} - \frac{3}{0.5} + \frac{2}{3}
$$

de unde obținem $y = 4.25$

## Exercițiul 11

În cerință primim un tabel cu valorile polinomului în anumite puncte.\
O să considerăm că acestea sunt polii de interpolare, deci avem
- $x_1 = 0, f(x_1) = 2$
- $x_2 = 1, f(x_2) = -1$
- $x_3 = 2, f(x_3) = 4$

Din faptul că toate diferențele divizate de ordinul 3 sunt $1$, rezultă că diferențele divizate de ordin mai mare vor fi $0$; deci polinomul $P_n$ este de **gradul 3**.

Ni se cere coeficientul lui $x^2$, acesta corespunde cu diferența divizată de ordin 2 $f[x_1, x_2, x_3]$.

Din definiția diferențelor divizate știm că
$$
f[x_1, x_2, x_3] = \frac{f[x_2, x_3] - f[x_1, x_2]}{x_3 - x_1}
$$
unde 
$$
\begin{align}
f[x_1, x_2] &= \frac{f(x_2) - f(x_1)}{x_2 - x_1} \\
f[x_2, x_3] &= \frac{f(x_3) - f(x_2)}{x_3 - x_2}
\end{align}
$$

Din datele pe care le avem, obținem
$$
\begin{align}
f[x_1, x_2] &= -3 \\
f[x_2, x_3] &= 5
\end{align}
$$
și
$$
f[x_1, x_2, x_3] = \frac{5 + 3}{2} = 4
$$

Deci coeficientul lui $x^2$ este $4$.

## Exercițiul 12

Folosind valorile date și formula de mai sus, avem polinomul:

$$
P_2(x) = f[x_1] + f[x_1, x_2] \cdot (x - 0) + f[x_1, x_2, x_3] \cdot (x - 0) \cdot (x - 0.4)
$$

## Exercițiul 13

Interpolare spline liniară - secțiunea V.1 din [cursul 10](https://drive.google.com/file/d/1qclFs2MCxV_heEU4toWulxuGy-dY5VTO/view)

### a)

Definiția poate fi găsită în curs.

### b)

Fie $f(x) = \ln x$. Avem punctele $x = (1, e, e^2)$ și $y = f(x) = (\ln 1, \ln e, \ln e^2) = (0, 1, 2)$.

Calculăm coeficienții:

$$
a_1 = f(x_1) = 0 \\
b_1 = \frac{f(x_2) - f(x_1)}{x_2 - x_1} = \frac{1 - 0}{e - 1} = \frac{1}{e - 1}
$$

<hr/>

$$
a_2 = f(x_2) = 1 \\
b_2 = \frac{f(x_3) - f(x_2)}{x_3 - x_2} = \frac{2 - 1}{e^2 - e} = \frac{1}{e \cdot (e - 1)}
$$

Cele două porțiuni sunt:

$$
S_1 \colon [1, e) \to \mathbb{R}, S_1(x) = a_1 + b_1 (x - 1) \\
S_2 \colon [e, e^2] \to \mathbb{R}, S_2(x) = a_2 + b_2 (x - e)
$$

Funcția spline liniară care interpolează $f$ pe $[1, e^2]$ este reuniunea funcțiilor definite mai sus.

## Exercițiul 14

Interpolare spline pătratică - secțiunea V.2 din [cursul 10](https://drive.google.com/file/d/1qclFs2MCxV_heEU4toWulxuGy-dY5VTO/view)

### a)

Definiția poate fi găsită în curs.

### b)

Avem aceeași funcție și aceleași puncte ca la _13. b)_

Coeficienții sunt:

$$
h_1 = x_2 - x_1 = e - 1 \\
a_1 = f(x_1) = 0 \\
b_1 = f'(x_1) = \frac{1}{e} \\
\begin{align}
c_1 &= \frac{1}{h_1^2} (f(x_2) - f(x_1) - h_1 b_1) \\
&= \frac{1}{{(e - 1)}^2} \left(1 - 0 - (e - 1) \cdot \frac{1}{e}\right) \\
&= \frac{1}{{(e - 1)}^2} - \frac{1}{(e - 1) \cdot e}
\end{align}
$$

<hr/>

$$
h_2 = x_3 - x_2 = e^2 - e \\
a_2 = f(x_2) = 1 \\
b_2 = \frac{2}{h_1} (f(x_2) - f(x_1)) - b_1 = \frac{2}{e - 1} (1 - 0) - \frac{1}{e} = \frac{2}{e-1} - \frac{1}{e} \\
\begin{align}
c_2 &= \frac{1}{{(e^2 - e)}^2} \left(2 - 1 - (e - 1) \cdot \left(\frac{1}{{(e - 1)}^2} - \frac{1}{(e - 1) \cdot e}\right)\right) \\
&= \frac{1}{{(e^2 - e)}^2} \left(1 - \frac{1}{e - 1} + \frac{1}{e} \right)
\end{align}
$$

Cele două porțiuni sunt $S_1(x) = a_1 + b_1 (x - 1) + c_1 (x - 1)^2$, $S_2(x) = a_2 + b_2 (x - e) + c_2 (x - e)^2$.

## Exercițiul 15

Ce se cere este să deducem/scriem schema numerică de calcul a coeficienților unei funcții de interpolare spline pătratice, care se găsește și în curs:

$$
a_i = f(x_i), i \in [1, n] \\
b_1 = f'(x_1) \\
b_i = \frac{2}{h_{i - 1}} \left(f(x_i) - f(x_{i-1}) \right) - b_{i-1}, i \in [2, n] \\
c_i = \frac{1}{h_i^2} \left(f(x_{i + 1} - f(x_i) - h_i b_i)\right), i \in [1, n]
$$

Este suficient să aplicăm formulele de mai sus pentru $i = 1$ și $i = 2$.

## Exercițiul 16

Funcția de interpolare spline cubică - [cursul 11](https://drive.google.com/file/d/1RYFk7vVf4D8IoGXkIPy1rZGiZTbt-PJH/view)

In [16]:
x = np.array([0, 1, 2], dtype=np.float64)
y = np.array([0, 1, 2], dtype=np.float64)

Formulele pentru calcularea coeficienților unei funcții spline cubice sunt mai complicate. Trebuie să definim un sistem de liniar și să îl rezolvăm pentru a putea determina $b$-urile, iar din acestea putem calcula $c$-urile și $d$-urile.

In [17]:
N = len(x) - 1

# Coeficienții `a` sunt egali cu `y`
a = y.copy()

B = np.zeros((N + 1, N + 1))

B[0, 0] = 1
for i in range(1, N):
    B[i, i - 1] = 1
    B[i, i] = 4
    B[i, i + 1] = 1
B[N, N] = 1

h = x[1] - x[0]

W = np.zeros((N + 1, 1))
for i in range(1, N):
    W[i] = 3 * (y[i + 1] - y[i - 1]) / h

# Coeficienții `b` se obțin rezolvând un sistem de ecuații (construit mai sus)
b = np.linalg.solve(B, W).reshape(N + 1)

c = np.zeros(N)
d = np.zeros(N)

h_squared = h * h
h_cubed = h * h * h

for i in range(N):
    # Coeficienții `c` și `d` se obțin din `b`
    c[i] = 3 * (y[i + 1] - y[i]) / h_squared - (b[i + 1] + 2 * b[i]) / h
    d[i] = - 2 * (y[i + 1] - y[i]) / h_cubed + (b[i + 1] + b[i]) / h_squared

print("a =", a[:-1])
print("b =", b[:-1])
print("c =", c)
print("d =", d)

a = [0. 1.]
b = [0.  1.5]
c = [1.5 0. ]
d = [-0.5 -0.5]


## Exercițiul 18

Din condiția că derivata de ordin I și II a funcției spline trebuie să fie continuă, avem că

$$
S_1'(1) = S_2'(1) \iff 2 - 3 \cdot 1^2 = B \iff \fbox{B = -1} \\
S_1''(1) = S_2''(1) \iff - 6 \cdot 1 = 2 C \iff \fbox{C = -3}
$$

Din ipoteză ni se spune și că $S''(b) = 0$, deci

$$
S_2''(2) = 0 \iff 2 C + 6 D (2 - 1) = 0 \iff 6 D - 6 = 0 \iff \fbox{D = 1}
$$

## Exercițiul 19

Scriem funcțiile $S_i$ și derivatele lor:

$$
\begin{align}
S_1 &\colon [1, 2) \to \mathbb{R} \\
S_1(x) &= 3(x - 1) + 2 (x - 1)^2 - (x - 1)^3 \\
S_1'(x) &= 3 + 4 (x - 1) - 3(x - 1)^2 \\
S_1''(x) &= 4 - 6 (x - 1)
\end{align}
$$

$$
\begin{align}
S_2 &\colon [2, 3] \to \mathbb{R} \\
S_2(x) &= A + B (x - 2) + C (x - 2)^2 + D (x - 2)^3 \\
S_2'(x) &= B + 2C (x - 2) + 3 D(x - 2)^2 \\
S_2''(x) &= 2C + 6D (x - 2)
\end{align}
$$

Știm că funcția $S$ trebuie să fie continuă, deci
$$
S_1(2) = S_2(2) \iff 3(2 - 1) + 2 (2 - 1)^2 - (2 - 1)^3 = A \iff 3 + 2 - 1 = A \iff \fbox{$A = 4$}
$$

Știm că funcția $S'$ trebuie să fie continuă, deci
$$
S_1'(2) = S_2'(2) \iff 3 + 4 (2 - 1) - 3 (2 - 1) = B \iff 3 + 4 - 3 = B \iff \fbox{$B = 4$}
$$

Știm că funcția $S''$ trebuie să fie continuă, deci
$$
S_1''(2) = S_2''(2) \iff 4 - 6 = 2 C \iff \fbox{$C = -1$}
$$

Din cerință ni se spune că
$$
f'(1) = f'(3) \iff S_1'(1) = S_2'(3) \iff 3 = 4 - 2 + 3D \iff \fbox{$D = \frac{1}{3}$}
$$

## Exercițiul 22

Derivarea numerică - [cursul 12](https://drive.google.com/file/d/1c6twFTnfwdq-3B9soATOsjv9NAh6811i/view)

Scriem [dezvoltarea în serie Taylor până la ordinul 3](https://www.wolframalpha.com/input/?i=Taylor+series+expansion+of+f%28x+%2B+h%29+up+to+order+3) pentru termeni:

$$
f(x - h) = f(x) - h f'(x) + \frac{h^2}{2} f''(x) - \frac{h^3}{6} f^{(3)}(x) + O(h^4) \\
f(x + h) = f(x) + h f'(x) + \frac{h^2}{2} f''(x) + \frac{h^3}{6} f^{(3)}(x) + O(h^4) \\
f(x + 2h) = f(x) + 2h f'(x) + 2 h^2 f''(x) + \frac{4 h^3}{3} f^{(3)}(x) + O(h^4) \\
f(x + 3h) = f(x) + 3h f'(x) + \frac{9 h^2}{2} f''(x) + \frac{9h^3}{2} f^{(3)}(x) + O(h^4)
$$

După aceea, scriem combinația liniară din indicație $A f(x - h) + B f(x + h) + C f(x + 2h) + D f(x + 3h)$ și obținem:

$$
f(x) \cdot (A + B + C + D) \\
+ h f'(x) \cdot (-A + B + 2C + 3D) \\
+ h^2 f''(x) \cdot \left(\frac{1}{2} A + \frac{1}{2} B + 2 C + \frac{9}{2} D\right) \\
+ h^3 f^{(3)}(x) \cdot \left(- \frac{1}{6} A + \frac{1}{6} B + \frac{4}{3} C + \frac{9}{2} D\right) + O(h^4)
$$

Vrem să obținem o formulă în care să rămână doar $f'(x)$, deci construim sistemul:
$$
\begin{pmatrix}
1 & 1 & 1 & 1 \\
-1 & 1 & 2 & 3 \\
\frac{1}{2} & \frac{1}{2} & 2 & \frac{9}{2} \\
-\frac{1}{6} & \frac{1}{6} & \frac{4}{3} & \frac{9}{2}
\end{pmatrix}
\begin{pmatrix}
A \\
B \\
C \\
D
\end{pmatrix}
=
\begin{pmatrix}
0 \\
1 \\
0 \\
0
\end{pmatrix}
$$

In [18]:
A = np.array([
    [1, 1, 1, 1],
    [-1, 1, 2, 3],
    [1/2, 1/2, 2, 9/2],
    [-1/6, 1/6, 4/3, 9/2]
])
b = np.array([[0, 1, 0, 0]]).T

np.linalg.solve(A, b)

array([[-0.45833333],
       [ 0.25      ],
       [ 0.33333333],
       [-0.125     ]])

Rezolvând sistemul obținem:

$$\frac{-0.4583 f(x - h) + 0.25 f(x + h) + 0.33 f(x + 2h) - 0.125 f(x + 3h)}{h} \approx f'(x)$$

cu ordinul de aproximare $\frac{O\left(h^4\right)}{h} = O\left(h^3\right)$

## Exercițiul 24

Putem să plecăm de la metoda de extrapolare Richardson descrisă în curs, dar în loc să folosim $\varphi(h), \varphi\left(\frac{h}{2}\right), \varphi\left(\frac{h}{2^2}\right)$ folosim $\varphi(h), \varphi\left(\frac{h}{3}\right), \varphi\left(\frac{h}{3^2}\right)$.

Pentru a păstra notația din curs, notăm
$$
Q_{i, j} = \varphi_j \left(\frac{h}{3^{i - j}}\right)
$$
definit inductiv prin
$$
Q_{i, j} = \varphi_{j - 1} \left(\frac{h}{3^{i - j + 1}}\right) + \frac{1}{3^{j - 1} - 1}\left(
\varphi_{j - 1}\left(\frac{h}{3^{i - (j - 1)}}\right) - \varphi_{j - 1}\left(\frac{h}{3^{i - 1 - (j - 1)}}\right)
\right)
$$

Scris mai pe scurt:
$$
Q_{i, j} = Q_{i, j-1} + \frac{1}{3^{j - 1} - 1} \left(Q_{i, j - 1} - Q_{i - 1, j - 1} \right)
$$
unde
$$
Q_{1, 1} = \varphi(h), Q_{2, 1} = \varphi\left(\frac{h}{3}\right), Q_{3, 1} = \varphi\left(\frac{h}{3^2}\right)
$$

Dacă aplicăm formula inductivă de mai sus plecând de la $Q_{3, 3}$, obținem aproximarea de ordinul $O(h^3)$ care se cere în enunț.

## Exercițiul 25

Formule de cuadratură Newton-Cotes - secțiunea VII.2 din [cursul 13](https://drive.google.com/file/d/1DCPEvUUGvzI1cGpAiGzLdMCv4qWPc4T3/view).

Formula de cuadratură a dreptunghiului este:
$$
I_0 (f) = (b - a) f\left(\frac{a + b}{2}\right)
$$

Formula de cuadratură a trapezului este:
$$
I_1 (f) = \frac{b - a}{2} \cdot \left[f(a) + f(b)\right]
$$

Formula de cuadratură Simpson este:
$$
I_2 (f) = \frac{b - a}{6} \cdot \left[f(a) + 4 f\left(\frac{a + b}{2}\right) + f(b)\right]
$$

În cazul nostru:
$$
\frac{2 - 0}{2} \cdot \left[f(0) + f(2)\right] = 5 \implies f(0) + f(2) = 5 \\
(2 - 0) \cdot f \left(\frac{0 + 2}{2}\right) = 2 f(1) = 4 \implies f(1) = 2
$$

Deci formula Simpson ne-ar da rezultatul
$$
I_2 (f) = \frac{2 - 0}{6} \cdot \left[f(0) + 4 f(1) + f(2)\right] = \frac{13}{3}
$$

## Exercițiul 27

Formule de cuadratură sumată - secțiunea VII.3 din [cursul 13](https://drive.google.com/file/d/1DCPEvUUGvzI1cGpAiGzLdMCv4qWPc4T3/view).

Formula de cuadratură sumată a trapezului este
$$
I_{1, m}(f) = \frac{h}{2}\left(f(x_1) + 2 \sum_{k = 2}^{m} f(x_k) + f(x_{m + 1}) \right)
$$

În cazul nostru, pentru $m = 4$, avem $h = \frac{1}{4} = 0.25$. Formula s-ar scrie ca:
$$
\begin{align}
I_{1, 4}(f) &= \frac{h}{2} (f(0) + 2 (f(0.25) + f(0.5) + f(0.75)) + f(1)) \\
&= 0.125 (1 + 2(\alpha + 2.5 + \alpha) + 2)) \\
&= 0.125 (8 + 4 \alpha) = 1 + 0.5 \alpha
\end{align}
$$
Egalând cu valoarea dată din cerință:
$$
1 + 0.5\alpha \approx 1.75 \iff 0.5 \alpha \approx 0.75 \iff \fbox{$\alpha \approx 1.5$}
$$